File: src/main/java/com/portfolio/saga/application/SagaController.java
package com.portfolio.saga.application;

import com.portfolio.saga.domain.OrderSaga;
import com.portfolio.saga.domain.Saga;
import com.portfolio.saga.domain.SagaOrchestrator;
import com.portfolio.saga.domain.SagaStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/api/saga")
public class SagaController {
    private final InMemorySagaLog sagaLog = new InMemorySagaLog();
    private final SagaOrchestrator orchestrator = new SagaOrchestrator();

    @PostMapping("/order")
    public ResponseEntity<Map<String, Object>> createOrder(@RequestBody Map<String, Object> body) {
        String orderId = body.getOrDefault("orderId", "order-" + System.currentTimeMillis()).toString();
        boolean failInventory = Boolean.parseBoolean(body.getOrDefault("failInventory", "false").toString());
        boolean failPayment = Boolean.parseBoolean(body.getOrDefault("failPayment", "false").toString());
        boolean failShip = Boolean.parseBoolean(body.getOrDefault("failShip", "false").toString());

        Saga saga = new OrderSaga(sagaLog, orderId, failInventory, failPayment, failShip);
        SagaStatus result = orchestrator.execute(saga);

        return ResponseEntity.ok(Map.of(
                "sagaId", saga.getId(),
                "orderId", orderId,
                "status", result.name()
        ));
    }

    @GetMapping("/health")
    public ResponseEntity<Map<String, String>> health() {
        return ResponseEntity.ok(Map.of("status", "UP"));
    }
}

File: src/main/kotlin/com/portfolio/saga/application/SagaController.kt
package com.portfolio.saga.application

import com.portfolio.saga.domain.FailurePlan
import com.portfolio.saga.domain.OrderStep
import com.portfolio.saga.domain.SagaInstance
import com.portfolio.saga.domain.SagaOrchestrator
import com.portfolio.saga.domain.SagaStore
import com.portfolio.saga.domain.SagaTransition
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.server.ResponseStatusException
import java.util.UUID

data class StartSagaRequest(
    val orderId: String,
    val failStep: OrderStep? = null,
    val failCompensationStep: OrderStep? = null,
    val crashAfterStep: OrderStep? = null,
)

data class SagaResponse(
    val saga: SagaInstance,
    val transitions: List<SagaTransition>,
)

@RestController
@RequestMapping("/api/sagas")
class SagaController(
    private val orchestrator: SagaOrchestrator,
    private val store: SagaStore,
) {
    @PostMapping
    fun start(@RequestBody request: StartSagaRequest): ResponseEntity<SagaResponse> {
        val result = orchestrator.start(
            request.orderId,
            FailurePlan(request.failStep, request.failCompensationStep, request.crashAfterStep),
        )
        return ResponseEntity.status(HttpStatus.CREATED).body(result.toResponse())
    }

    @PostMapping("/{sagaId}/recover")
    fun recover(@PathVariable sagaId: UUID): SagaResponse = orchestrator.resume(sagaId).toResponse()

    @PostMapping("/recover")
    fun recoverAll(): List<SagaResponse> = orchestrator.recover().map { it.toResponse() }

    @GetMapping("/{sagaId}")
    fun get(@PathVariable sagaId: UUID): SagaResponse =
        (store.get(sagaId) ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)).toResponse()

    private fun SagaInstance.toResponse() = SagaResponse(this, store.transitions(id))
}
